crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1628 @@
+            let mut current_shared_file_state: Arc<tokio::sync::OnceCell<SharedFileState>> =
+                Default::default();
+            let mut current_path_index = 0;
+
+            loop {
+                let semaphore_permit = semaphore.clone().acquire_owned().await.unwrap();
+
+                let tx_permit = {
+                    use tokio::sync::mpsc::error::*;
+                    match row_group_tx.try_reserve() {
+                        Err(TrySendError::Closed(_)) => break,
+                        Ok(v) => v,
+                        // Should not fail: We already have the semaphore permit.
+                        v @ Err(TrySendError::Full(_)) => v.unwrap(),
+                    }
+                };
+
+                let Some((path_index, v)) = row_group_data_rx.recv().await else {
+                    break;
+                };
+
+                if path_index != current_path_index {
+                    current_shared_file_state = Default::default();
+                }
+
+                current_path_index = path_index;
+
+                let row_group_decoder = row_group_decoder.clone();
+                let current_shared_file_state = current_shared_file_state.clone();
+                let handle = cpu_runtime.spawn_task(TaskPriority::Low, async move {
+                    // Hold the permit until we've finished fetching and decoding the row group.
+                    let (handle, _row_group_data_fetch_permit) = WithSemaphorePermit::split(v);
+                    let row_group_data = handle.await?;
+
+                    row_group_decoder
+                        .row_group_data_to_df(row_group_data, current_shared_file_state.as_ref())
+                        .await
+                });
+
+                let handle = async_executor::AbortOnDropHandle::new(handle);
+                let handle = WithSemaphorePermit::new(handle, semaphore_permit);
+
+                tx_permit.send(handle);
+            }
+
+            // Join on the producer handle to catch errors/panics.
+            // Safety: The receiver is dropped to break the producer loop before joining.
+            drop(row_group_data_rx);
+            row_group_data_task_handle.await
+        });
+
+        (
+            row_group_rx,
+            task_handles_ext::AbortOnDropHandle(row_group_task_handle),
+        )
+    }
+
+    /// Constructs the task that fetches the bytes of row groups.
+    #[allow(clippy::type_complexity)]
+    fn init_row_group_data_fetcher(
+        &mut self,
+    ) -> (
+        tokio::sync::mpsc::Receiver<(
+            usize,
+            WithSemaphorePermit<async_executor::AbortOnDropHandle<PolarsResult<RowGroupData>>>,
+        )>,
+        async_executor::AbortOnDropHandle<PolarsResult<()>>,
+    ) {
+        let verbose = self.verbose;
+        let io_runtime = polars_io::pl_async::get_runtime();
+
+        let row_group_prefetch_size = self.config.row_group_prefetch_size;
```

Comment:
`POLARS_ROW_GROUP_PREFETCH_SIZE` currently defaults to 128 - for the existing parquet streaming node the row group prefetch is also limited to 1 file, but this parquet source node pre-fetches row groups potentially across files to reach the prefetch size.